Gems/ROS2/Code/Source/Frame/ROS2FrameSystemComponent.cpp

```diff
@@ -0,0 +1,440 @@
+        }
+
+        return false;
+    }
+
+    AZ::EntityId ROS2FrameSystemComponent::GetParentEntityId(const AZ::EntityId& frameEntityId) const
+    {
+        if (m_frameParent.contains(frameEntityId))
+        {
+            return m_frameParent.find(frameEntityId)->second;
+        }
+        return AZ::EntityId();
+    }
+
+    AZStd::vector<AZ::EntityId> ROS2FrameSystemComponent::FindFrameParentPath(AZ::EntityId frameEntityId)
+    {
+        AZStd::vector<AZ::EntityId> path;
+        path.push_back(frameEntityId);
+
+        AZ::Entity* currentEntity = nullptr;
+        AZ::ComponentApplicationBus::BroadcastResult(currentEntity, &AZ::ComponentApplicationRequests::FindEntity, frameEntityId);
+        auto* currentTransform = GetEntityTransformInterface(currentEntity);
+        if (!currentTransform)
+        {
+            return path;
+        }
+        while (currentEntity->GetId().IsValid())
+        {
+            AZ::EntityId nextEntityId;
+            if (!currentTransform)
+            {
+                return path;
+            }
+            nextEntityId = currentTransform->GetParentId();
+            AZ::Entity* parent = nullptr;
+            AZ::ComponentApplicationBus::BroadcastResult(parent, &AZ::ComponentApplicationRequests::FindEntity, nextEntityId);
+            if (parent == nullptr || !nextEntityId.IsValid())
+            { // Found top of the level
+                path.push_back(nextEntityId);
+                return path;
+            }
+            if (ROS2FrameComponentBus::HasHandlers(nextEntityId))
+            {
+                path.push_back(nextEntityId);
+                break;
+            }
+            path.push_back(nextEntityId);
+            currentEntity = parent;
+            currentTransform = GetEntityTransformInterface(currentEntity);
+        }
+
+        return path;
+    }
+
+    void ROS2FrameSystemComponent::RegisterFrame(const AZ::EntityId& frameToRegister)
+    {
+        // Check if the frame is valid and if it's already registered;
+        if (!frameToRegister.IsValid() || m_frameParent.contains(frameToRegister))
+        {
+            return;
+        }
+
+        // Get path from itself to its parent (both included)
+        AZStd::vector<AZ::EntityId> entityPath = FindFrameParentPath(frameToRegister);
+        const AZ::EntityId& frameParent = entityPath[entityPath.size() - 1];
+
+        // Initialize all maps
+        AZStd::set<AZ::EntityId>& frameToRegisterChildren = m_frameChildren.insert({ frameToRegister, {} }).first->second;
+        AZStd::set<AZ::EntityId>& frameToRegisterWatchedEntities = m_watchedEntities.insert({ frameToRegister, {} }).first->second;
+        m_frameParent.insert({ frameToRegister, frameParent });
+
+        bool laysOnPath = false;
+
+        if (m_frameChildren.contains(frameParent))
+        {
+            AZStd::set<AZ::EntityId>& parentsChildren = m_frameChildren.find(frameParent)->second;
```

Comment:
.find() can be merge with .contains inside if-statement in line above + introducing != m_frameChildren.end() check -> this would result in one less m_frameChildren search.